### DISCUSSION TITLE
Idea: scope_exit

### DISCUSSION THREAD

**Discussion by pfeatherstone** [davisking/dlib#2874] [Ideas]
This might be a nice PR:
https://godbolt.org/z/zocG4ejhs

The non-erased version is a lot better but there are use-cases for the erased version.
Also note, the erased version is slightly more efficient than using `std::shared_ptr<void>` with a custom deleter which calls exit function.

**Comment by davisking** [comment]
Nice, yeah, would be cool to have this kind of thing on hand.  Have this tool at work and it's useful :)

**Comment by pfeatherstone** [comment]
Added https://github.com/davisking/dlib/pull/2875
